packages/core/src/stateUtils.ts

```diff
@@ -1823,3 +1817,51 @@ export function resolveStateValue(
   const allStateNodes = getAllStateNodes(getStateNodes(rootNode, stateValue));
   return getStateValue(rootNode, [...allStateNodes]);
 }
+
+/**
+ * Runs an executable action. Executable actions are returned from the
+ * `transition(…)` function.
+ *
+ * @example
+ *
+ * ```ts
+ * const [state, actions] = transition(someMachine, someState, someEvent);
+ *
+ * for (const action of actions) {
+ *   // Executes the action
+ *   executeAction(action);
+ * }
+ * ```
+ */
+export function executeAction(
+  action: ExecutableActionObject,
+  actor: AnyActorRef = createEmptyActor()
+) {
+  const actorScope = (actor as any)._actorScope as AnyActorScope;
+  const defer = actorScope.defer;
+  actorScope.defer = (fn) => fn();
+  try {
+    switch (action.type) {
+      case 'xstate.raise':
+        if (typeof (action as any).params.delay !== 'number') {
+          return;
+        }
+        executeRaise(actorScope, action.params as any);
+        return;
+      case 'xstate.sendTo':
+        executeSendTo(actorScope, action.params as any);
+        return;
+    }
+
+    action.exec?.(
+      {
+        ...action.info,
+        self: actor,
+        system: actor.system
+      },
+      action.params
+    );
```

Comment:
@davidkpiano this is missing action lookup in `machine.implementations.actions`. You can't rely on `.exec` being available if the goal is to have serializable actions. In fact, I'd probably just forbid combining inline actions with `machine.executeAction`. It should be an invariant that to use this API all actions must be declared explicitly in `machine.implementations.actions`. It's impossible to resolve an enqueued inline action, they are not discoverable in the static config at all. So it would be best to forbid all inline actions to make the rules clear around this.

Reply:
We fixed this, right?